Validators Testnet Docs (#190)
helium/docs: docs/mine-hnt/validators/setup.mdx

---
id: validator-run

> title: Validators (Testnet)

Reviewer: This should probably be Run a Validator

# Run a validator

The following options exist to build a Validator node:

* From [docker image](#docker-image)
* From [source](#source)

Regardless of the option, the genesis block needs to be installed:

```
Download command goes here
miner genesis load </absolute/path/to/genesis/block>
```

## Configure cloud platform for initial blockhain sync
If you choose to run your validator in a cloud platform (e.g., Amazon AWS) your CPU usage credits could easily be depleted during the initial sync of the blockchain. To avoid having your instance throttled (e.g., Amazon EC2 instances), check with your cloud platform to see if you can temporarily uncap your instance by setting CPU credit usage to unlimited.

## Port forwarding

Regardless of using docker or source, before launching a validator node configure the TCP port on your network. Validators communicate to peers over port: 2154/TCP. The networking logic knows how to get around a lack of forwarding, but validators will gain improved performance by forwarding the port.

## <a name="docker-image"></a>From docker image

See technical requirements [here](/mine-hnt/validators/requirements).

> First, update the package manager registry:

Reviewer: Probably best to put a new line after this line and before the update command.

```sudo apt-get update```

Then, install Docker:

```sudo apt-get install docker.io```

To avoid needing to use docker with sudo privileges, add your user to the docker group, replacing ```$USER``` with your username:

```sudo usermod -aG docker $USER```

Log in and out of your account to apply these changes. You are now ready to use Docker.

### Run a docker container

Note: on amd64 systems AVX support is required. Verify that it exists on your host system:
```
grep avx /proc/cpuinfo
grep avx2 /proc/cpuinfo
```
If nothing is returned from these commands, your host system does not have AVX support and your validator may not be stable.

### Start container

Before running the container for the first time, it is advisable to pick a "system mount point", i.e., a directory in your host system; some long-term data is stored there. This allows you to easily maintain your validator's blockchain identity (i.e. swarm keys) and blockchain state through image updates.

If you are using a Linux system, you can just create a directory in your user's home directory:

> ```mkdir ~/miner_data```

Reviewer 2: ```suggestion
```mkdir ~/validator_data```
```

If you are using Ubuntu as user ubuntu, this path would now be ```/home/ubuntu/miner_data```. This will be used later.
Now you can try the ```run``` command to start your container for the first time:

```
docker run -d \
--restart always \
--publish 2154:2154/tcp \

> --name miner \

Reviewer 2: ```suggestion
--name validator \
```

> --mount type=bind,source=/home/ubuntu/miner_data,target=/var/data \

Reviewer 2: ```suggestion
--mount type=bind,source=/home/ec2-user/validator_data,target=/var/data \
```

> quay.io/team-helium/miner:miner-xxxNN_YYYY.MM.DD

Reviewer 2: ```suggestion
quay.io/team-helium/validator:miner-val-{arch}_val{version}
```

```

> Replace ```xxxNN``` with the architecture used, i.e. amd64 or arm64, and with the release date of the image.

Reviewer 2: ```suggestion
Replace`{arch}` with the architecture used, i.e. `amd64` or `arm64`, and `{version}` with the version number of the image. We advise not using the latest tags as the underlying container will not be consistent.
```

The ```-d``` option runs in detached mode, which makes the command return or not; you may want to omit if you have a daemon manager running the docker for you.

The ```--restart always``` option asks Docker to keep the image running, starting the image on boot and restarting the image if it crashes. Depending on how you installed Docker in your system, it'll start on boot. In the AWS AMI above, we use systemd (```systemctl status docker``` to check).

> The ```--name miner``` names the container, which makes interacting with the docker easier, but feel free to name the container whatever you want.

Reviewer 2: ```suggestion
The ```--name validator``` names the container, which makes interacting with the docker easier, but feel free to name the container whatever you want.
```

> The ```--mount``` with the parameters above will mount the container's ```/var/data/``` directory to the systems directory ```/home/ec2-user/miner_data```.

Reviewer 2: ```suggestion
The ```--mount``` with the parameters above will mount the container's ```/var/data/``` directory to the systems directory ```/home/ec2-user/validator_data```. You will want this folder to be persistent across runs of the docker container as it will contain both the blockchain data and the miner key of your validator. 
```

### Interact with the validator within the container

You may want to interrogate the Validator or interact with it. Docker's ```exec``` command enables this e.g.:
```docker exec miner miner info height```

In other words, prepend ```docker exec miner``` to any of the commands documented or create an alias such as: ```alias miner="docker exec miner miner"```.
And start the container again as described above, but with the new release tag. 

Thanks to the ```--mount``` option, the blockchain data and the validator keys are preserved through updates.

After setup is complete, check [Monitor a validator section](/mine-hnt/validators/validator-monitor).

> ## <a name="source"></a>From source

Reviewer 2: I think @amirhaleem was going to take a stab here but I can come back to it if needed.

Author: you mean testing source e2e

---
id: validator-run
title: Run a Validator
hide_title: true
sidebar_label: Run a Validator
slug: /mine-hnt/validators/validator-run
---

import useBaseUrl from "@docusaurus/useBaseUrl";

# Run a Validator

The following options exist to build a validator node:

* From [docker image](#docker-image)
* From [source](#source)

Regardless of the option, the genesis block needs to be installed:

```

> Download command goes here

Reviewer 3: Seems like we're missing a command here?

Author: yes, updated...

miner genesis load </absolute/path/to/genesis/block>
```

## Configure cloud platform for initial blockhain sync
If you choose to run your validator in a cloud platform (e.g., Amazon AWS) your CPU usage credits could easily be depleted during the initial sync of the blockchain. To avoid having your instance throttled (e.g., Amazon EC2 instances), check with your cloud platform to see if you can temporarily uncap your instance by setting CPU credit usage to unlimited.

> ## Port forwarding

Reviewer 3: ```suggestion
## Port Forwarding
```

Regardless of using docker or source, before launching a validator node configure the TCP port on your network. Validators communicate to peers over port: 2154/TCP. The networking logic knows how to get around a lack of forwarding, but validators will gain improved performance by forwarding the port.

## <a name="docker-image"></a>From Docker image

See technical requirements [here](/mine-hnt/validators/requirements).

### Install Docker

First, update the package manager registry:

```sudo apt-get update```

Then, install Docker:

```sudo apt-get install docker.io```

To avoid needing to use docker with sudo privileges, add your user to the docker group, replacing ```$USER``` with your username:

```sudo usermod -aG docker $USER```

Log in and out of your account to apply these changes. You are now ready to use Docker.

Note: on amd64 systems AVX support is required. Verify that it exists on your host system:
```
grep avx /proc/cpuinfo
grep avx2 /proc/cpuinfo
```
If nothing is returned from these commands, your host system does not have AVX support and your validator may not be stable.

> ### Run a container

Reviewer 3: ```suggestion
### Run a Container
```

Then, install Docker:

```sudo apt-get install docker.io```

To avoid needing to use docker with sudo privileges, add your user to the docker group, replacing ```$USER``` with your username:

```sudo usermod -aG docker $USER```

Log in and out of your account to apply these changes. You are now ready to use Docker.

Note: on amd64 systems AVX support is required. Verify that it exists on your host system:
```
grep avx /proc/cpuinfo
grep avx2 /proc/cpuinfo
```
If nothing is returned from these commands, your host system does not have AVX support and your validator may not be stable.

### Run a container

Before running the container for the first time, it is advisable to pick a "system mount point". This allows you to easily maintain your validator's blockchain identity (i.e. swarm keys) and blockchain state through image updates. 

The validator directory should be created on a persistent ebs volume (if aws) or something similar that lives across server lifetimes.

```mkdir $HOME/validator_data; source=$HOME/validator_data```

Use the ```run``` command to start your container for the first time:

```
docker run -d \
--restart always \
--publish 2154:2154/tcp \
--name validator \

> --mount type=bind,source=/home/ubuntu/validator_data,target=/var/data \

Reviewer 4: maybe use the same `$HOME` env var here rather than assuming ubuntu?

Author: updated

Then, install Docker:

```sudo apt-get install docker.io```

To avoid needing to use docker with sudo privileges, add your user to the docker group, replacing ```$USER``` with your username:

```sudo usermod -aG docker $USER```

Log in and out of your account to apply these changes. You are now ready to use Docker.

Note: on amd64 systems AVX support is required. Verify that it exists on your host system:
```
grep avx /proc/cpuinfo
grep avx2 /proc/cpuinfo
```
If nothing is returned from these commands, your host system does not have AVX support and your validator may not be stable.

### Run a Container

Before running the container for the first time, it is advisable to pick a "system mount point". This allows you to easily maintain your validator's blockchain identity (i.e. swarm keys) and blockchain state through image updates. 

The validator directory should be created on a persistent ebs volume (if aws) or something similar that lives across server lifetimes.

```mkdir $HOME/validator_data; source=$HOME/validator_data```

Use the ```run``` command to start your container for the first time:

```
docker run -d \
--restart always \
--publish 2154:2154/tcp \
--name validator \

> --mount type=bind,source=$HOME/validator_data/validator_data,target=/var/data \

Reviewer 5: `source=$HOME/validator_data/validator_data`
should be
`source=$HOME/validator_data`

Author: fixed, thanks
